pretix_wallet/signals.py

```diff
@@ -12,19 +20,57 @@ def register_payment_provider(sender, **kwargs):
     return [WalletPaymentProvider]
 
 
+@receiver(checkout_flow_steps, dispatch_uid="checkout_flow_steps_wallet")
+def wallet_checkout_flow_steps(sender, **kwargs):
+    return MembershipGrantingCheckoutFlowStep
+
+
 @receiver(order_paid, dispatch_uid="payment_wallet_order_paid")
 def wallet_order_paid(sender, order, **kwargs):
-    top_up_positions = list(filter(lambda pos: position_is_top_up_product(pos), order.positions))
+    top_up_positions = list(filter(lambda pos: position_is_top_up_product(sender, pos), order.positions.all()))
     if top_up_positions:
         CustomerWallet.create_if_non_existent(sender.organizer, order.customer)
         try:
-            top_up_value = sum(lambda pos: pos.price, top_up_positions)
+            top_up_value = sum(map(lambda pos: pos.price, top_up_positions))
             _wallet_transaction(sender, order.payments.last(), order.customer.wallet.giftcard, sign=1,
                                 amount=top_up_value)
         except PaymentException as e:
             raise e
 
 
+class MembershipGrantingCheckoutFlowStep(CartMixin, BaseCheckoutFlowStep):
+    icon = 'user-plus'
+    identifier = 'wallet-membership-granting'
+
+    def label(self):
+        return _('Creating Wallet')
+
+    @cached_property
+    def priority(self):
+        # One less than MembershipStep
+        return 46
+
+    def get(self, request):
+        if request.event.organizer and request.customer:
+            CustomerWallet.create_if_non_existent(request.event.organizer, request.customer)
+            return redirect_to_url(self.get_next_url(request))
+        else:
+            raise ImproperlyConfigured('User reached the wallet creation step without signing in.'
+                                       'Have you created customer accounts and required membership'
+                                       'for the top-up product?')
+
+    def is_completed(self, request, warn=False):
+        if self.request.customer.wallet:
```

Comment:
IIRC this raises if not existent

Reply:
Yes, this should never happen (see  `get`). This I intended this to raise as this ever happening is a bug. Then, some of our central assumptions must be pretty wrong.